compiler/AST/foralls.cpp

```diff
@@ -230,3 +235,495 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+  return result;
+}
+
+static void buildLeaderLoopBody(ForallStmt* pfs, Expr* iterExpr) {
+  bool zippered = pfs->zippered();
+
+  // Set up the follower call etc.
+
+  // Extract leadIdxCopy and its type.
+  VarSymbol* leadIdxCopy     = parIdxCopyVar(pfs);
+
+  BlockStmt* resultBlock     = new BlockStmt();
+  // cf in build.cpp: new ForLoop(leadIdx, leadIter, NULL, zippered)
+  BlockStmt* leadForLoop     = new BlockStmt();
+
+  VarSymbol* iterRec         = newTemp("chpl__iterLF"); // serial iter, LF case
+
+  VarSymbol* followIdx       = newTemp("chpl__followIdx");
+  VarSymbol* followIter      = newTemp("chpl__followIter");
+  BlockStmt* followBlock     = NULL;
+
+  iterRec->addFlag(FLAG_EXPR_TEMP);
+  iterRec->addFlag(FLAG_CHPL__ITER);
+
+  followIdx->addFlag(FLAG_INDEX_OF_INTEREST);
+
+  resultBlock->insertAtTail(new DefExpr(iterRec));
+  resultBlock->insertAtTail(new CallExpr(PRIM_MOVE, iterRec, iterExpr));
+  Expr* toNormalize = resultBlock->body.tail;
+
+  // loop body to refer to followIdx instead of leadIdxCopy
+  for_SymbolSymExprs(se, leadIdxCopy)
+    if (se->parentExpr->parentExpr != pfs->loopBody()) // heuristic
+      se->replace(new SymExpr(followIdx));
+
+  // todo rename loopBody to userBody
+  BlockStmt* loopBody = userLoop(pfs);
+  loopBody->remove();
+
+  followBlock = buildFollowLoop(iterRec,
+                                leadIdxCopy,
+                                followIter,
+                                followIdx,
+                                loopBody,
+                                pfs,
+                                false,
+                                zippered);
+
+  if (fNoFastFollowers == false) {
+    // from the original buildForallLoopStmt()
+    Symbol* T1 = newTemp();
+    Symbol* T2 = newTemp();
+
+    VarSymbol* fastFollowIdx   = newTemp("chpl__fastFollowIdx");
+    VarSymbol* fastFollowIter  = newTemp("chpl__fastFollowIter");
+    BlockStmt* fastFollowBlock = NULL;
+
+
+    T1->addFlag(FLAG_EXPR_TEMP);
+    T1->addFlag(FLAG_MAYBE_PARAM);
+
+    T2->addFlag(FLAG_EXPR_TEMP);
+    T2->addFlag(FLAG_MAYBE_PARAM);
+
+    leadForLoop->insertAtTail(new DefExpr(T1));
+    leadForLoop->insertAtTail(new DefExpr(T2));
+
+    if (zippered == false) {
+      leadForLoop->insertAtTail("'move'(%S, chpl__staticFastFollowCheck(%S))",    T1, iterRec);
+      leadForLoop->insertAtTail(new CondStmt(new SymExpr(T1),
+                                          new_Expr("'move'(%S, chpl__dynamicFastFollowCheck(%S))",    T2, iterRec),
+                                          new_Expr("'move'(%S, %S)", T2, gFalse)));
+    } else {
+      leadForLoop->insertAtTail("'move'(%S, chpl__staticFastFollowCheckZip(%S))", T1, iterRec);
+      leadForLoop->insertAtTail(new CondStmt(new SymExpr(T1),
+                                          new_Expr("'move'(%S, chpl__dynamicFastFollowCheckZip(%S))", T2, iterRec),
+                                          new_Expr("'move'(%S, %S)", T2, gFalse)));
```

Comment:
I'd probably write this code as setting two `const char*` variables based upon `zippered` to name the function to call, since the rest is identical. I don't particularly care for the `new_Expr` thing but I don't object to you using it. In this case though, it might be interfering with writing it the way I would have.